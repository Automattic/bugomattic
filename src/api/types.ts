import { LearnMoreLink, TaskMapping } from '../reporting-config';

/**
 * An interface representing an interaction with the API.
 * Can create multiple implementations to support local testing and production deployment.
 */
export interface ApiClient {
	loadReportingConfig(): Promise< ReportingConfigApiResponse >;
}

/**
 * The issue reporting config as stored in and fetched from the database.
 * It follows a more heirarchical structure for easy reading and manual updating.
 */
export interface ReportingConfigApiResponse {
	[ productName: string ]: Product;
}

interface Product {
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
	featureGroups?: FeatureGroups;
	features?: Features;
}

interface FeatureGroups {
	[ featureGroupName: string ]: FeatureGroup;
}

interface FeatureGroup {
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
	features: Features;
}

interface Features {
	[ featureName: string ]: Feature;
}

interface Feature {
	keywords?: string[];
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
}
