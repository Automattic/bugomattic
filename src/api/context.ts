import { createContext } from 'react';
import { ApiClient } from './types';

export const ApiContext = createContext< ApiClient | undefined >( undefined );
