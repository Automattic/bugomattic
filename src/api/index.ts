export * from './types';
export * from './local';
export * from './use';
