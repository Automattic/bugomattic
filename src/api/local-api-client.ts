import { ApiClient } from './types';
import path from 'path-browserify';
import { Issue } from '../duplicate-results/types';

/**
 * An implementation of the API client that returns faked values for local development.
 */
export const localApiClient: ApiClient = {
	loadReportingConfig: async () => {
		// Let's keep us honest by adding a little delay here.
		await wait( 2000 );

		const reportingConfigName = path.parse(
			process.env.REACT_APP_REPORTING_CONFIG_NAME || 'default'
		).name;
		const reportingConfigUrl = `${ process.env.PUBLIC_URL }/local-reporting-configs/${ reportingConfigName }.json`;

		try {
			const reportingConfigRepsonse = await fetch( reportingConfigUrl );
			return reportingConfigRepsonse.json();
		} catch ( error ) {
			throw new Error(
				`Unable to find and parse local reporting config called ${ reportingConfigName }.json. ` +
					`Use the env variable "REACT_APP_REPORTING_CONFIG_NAME" and ensure there is a JSON file with a matching name in the "public/local-reporting-configs" directory.` +
					`Original error: ${ error }`
			);
		}
	},

	searchIssues: async ( search, options ) => {
		// Delay to simulate network latency.
		await wait( 2000 );

		// We can tweak this as needed during testing!
		// As a baseline, we'll show all the passed search parameters in the content.
		// And we'll randomize some other stuff so there's good variability.

		const numberOfIssues = Math.floor( Math.random() * 20 );
		const issues: Issue[] = [];
		for ( let i = 0; i < numberOfIssues; i++ ) {
			const randomString = Math.random().toString( 16 ).slice( 2 );
			const title = `Issue ${ randomString }`;

			const providedRepos = ( options?.repos || [ 'none provided' ] ).join( ', ' );
			const providedStatus = options?.status || 'none provided';
			const providedSort = options?.sort || 'none provided';
			const content = `Search: <em data-search-match>${ search }</em>. And another match: <em data-search-match>foo</em>. | Repos: ${ providedRepos } | Status: ${ providedStatus } | Sort: ${ providedSort }`;

			const status = Math.random() > 0.5 ? 'open' : 'closed';

			const oneWeekInMilliseconds = 7 * 24 * 60 * 60 * 1000;
			const randomDate = new Date(
				Date.now() - Math.floor( Math.random() * oneWeekInMilliseconds )
			).toISOString();

			issues.push( {
				dateCreated: randomDate,
				dateUpdated: randomDate,
				title,
				content,
				status,
				author: 'Fake Author',
				repo: 'Fakeorg/fake-repo',
				url: `https://github.com/Automattic/wp-calypso/issues/${ i }`,
			} );
		}

		return issues;
	},

	loadAvailableRepoFilters: async () => {
		await wait( 1000 );
		return [
			'Automattic/wp-calypso',
			'Automattic/wp-desktop',
			'Automattic/jetpack',
			'Automattic/themes',
			'Automattic/simplenote-ios',
			'Automattic/simplenote-android',
			'Automattic/pocket-casts-ios',
			'Automattic/sensei',
			'Automattic/woocommerce.com',
			'Automattic/newspack-blocks',
			'wordpress-mobile/WordPress-iOS',
			'wordpress-mobile/WordPress-Android',
			'woocommerce/woocommerce',
			'woocommerce/woocommerce-blocks',
			'woocommerce/woocommerce-admin',
			'woocommerce/woocommerce-ios',
			'woocommerce/woocommerce-android',
		];
	},
};

async function wait( ms: number ) {
	return new Promise( ( resolve ) => {
		setTimeout( resolve, ms );
	} );
}
