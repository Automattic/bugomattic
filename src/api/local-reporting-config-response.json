{
	"Product": {
		"description": "Product description",
		"learnMoreLinks": [
			{
				"type": "p2",
				"subdomain": "product"
			},
			{
				"type": "slack",
				"channel": "product"
			},
			{
				"type": "general",
				"href": "https://google.com/search?q=product"
			}
		],
		"tasks": {
			"bug": [
				{
					"instructions": "Product | Bug | GitHub link",
					"link": {
						"type": "github",
						"repository": "Automattic/fake-repo",
						"labels": [ "[Type] Bug" ]
					}
				}
			],
			"featureRequest": [
				{
					"instructions": "Product | Feature Request | P2 link",
					"link": {
						"type": "p2",
						"subdomain": "product"
					}
				}
			],
			"blocker": [
				{
					"instructions": "Product | Blocker | Slack link",
					"link": {
						"type": "slack",
						"channel": "product"
					}
				}
			]
		},
		"featureGroups": {
			"Feature Group": {
				"description": "Feature Group description",
				"learnMoreLinks": [
					{
						"type": "p2",
						"subdomain": "feature-group"
					},
					{
						"type": "slack",
						"channel": "feature-group"
					},
					{
						"type": "general",
						"href": "https://google.com/search?q=feature-group"
					}
				],
				"tasks": {
					"bug": [
						{
							"instructions": "Feature Group | Bug | Instructions only"
						}
					],
					"featureRequest": [
						{
							"instructions": "Feature Group | Feature Request | GitHub link",
							"link": {
								"type": "github",
								"repository": "Automattic/fake-repo",
								"labels": [ "[Type] Feature Request" ]
							}
						}
					],
					"blocker": [
						{
							"instructions": "Duplicate Task | P2 link",
							"link": {
								"type": "p2",
								"subdomain": "duplicate"
							}
						}
					]
				},
				"features": {
					"Feature Under Feature Group": {
						"description": "Feature Under Feature Group description",
						"keywords": [ "foo", "bar" ],
						"learnMoreLinks": [
							{
								"type": "p2",
								"subdomain": "feature-under-feature-group"
							},
							{
								"type": "slack",
								"channel": "feature-under-feature-group"
							},
							{
								"type": "general",
								"href": "https://google.com/search?q=feature-under-feature-group"
							}
						],
						"tasks": {
							"bug": [
								{
									"instructions": "Feature Under Feature Group | Bug | GitHub link",
									"link": {
										"type": "github",
										"repository": "Automattic/fake-repo",
										"labels": [ "[Type] Bug" ],
										"projectSlugs": [ "[Automattic/12345]" ]
									}
								}
							],
							"featureRequest": [
								{
									"instructions": "Feature Under Feature Group | Feature Request | GitHub link",
									"link": {
										"type": "github",
										"repository": "Automattic/fake-repo",
										"labels": [ "[Type] Feature Request" ],
										"projectSlugs": [ "[Automattic/12345]" ]
									}
								}
							],
							"blocker": [
								{
									"instructions": "Feature Under Feature Group | Blocker | GitHub link",
									"link": {
										"type": "github",
										"repository": "Automattic/fake-repo",
										"labels": [ "[Type] Bug", "[Pri] BLOCKER" ],
										"projectSlugs": [ "[Automattic/12345]" ]
									}
								},
								{
									"instructions": "Duplicate Task | P2 link",
									"link": {
										"type": "p2",
										"subdomain": "duplicate"
									}
								}
							]
						}
					},
					"Feature Directly Under Product": {
						"description": "Feature Directly Under Product description (has no tasks of its own)",
						"keywords": [ "baz", "qux" ],
						"learnMoreLinks": [
							{
								"type": "p2",
								"subdomain": "feature-directly-under-product"
							},
							{
								"type": "slack",
								"channel": "feature-directly-under-product"
							},
							{
								"type": "general",
								"href": "https://google.com/search?q=feature-under-feature-group"
							}
						]
					}
				}
			}
		}
	},
	"Alternative Product": {
		"description": "Alternative Product (providers more bare-bones entities to search on)",
		"featureGroups": {
			"Alternative Feature Group": {
				"features": {
					"Alternative Feature": {
						"tasks": {
							"bug": [
								{
									"instructions": "Alternative Feature | Bug | Instructions only"
								}
							],
							"featureRequest": [
								{
									"instructions": "Alternative Feature | Feature Request | Instructions only"
								}
							],
							"blocker": [
								{
									"instructions": "Alternative Feature | Blocker | Instructions only"
								}
							]
						}
					}
				}
			}
		}
	}
}
