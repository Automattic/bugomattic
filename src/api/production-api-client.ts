import { LoggerApiClient, LogPayload } from '../monitoring/types';
import {
	ApiClient,
	IssueApiResponse,
	ReportingConfigApiResponse,
	SearchIssueOptions,
} from './types';

class ProductionApiClient implements ApiClient, LoggerApiClient {
	private nonce: string;
	private nonceHeaderName: string;

	constructor() {
		this.nonce = globalThis.nonce;
		this.nonceHeaderName = globalThis.nonceHeaderName;
	}

	async loadReportingConfig(): Promise< ReportingConfigApiResponse > {
		const cachedData = localStorage.getItem( 'cachedReportingConfigData' );
		const cacheExpiry = localStorage.getItem( 'cacheExpiry' );
		if ( cachedData && cacheExpiry && Date.now() < +cacheExpiry ) {
			return JSON.parse( cachedData );
		}

		const request = new Request( '/wp-json/bugomattic/v1/reporting-config/', {
			method: 'GET',
			credentials: 'same-origin',
			headers: new Headers( {
				[ this.nonceHeaderName ]: this.nonce,
			} ),
		} );
		const response = await fetch( request );

		if ( response.ok ) {
			const data = await response.json();

			const cacheExpiryTime = Date.now() + 3 * 24 * 60 * 60 * 1000; // 3 days in milliseconds
			localStorage.setItem( 'cachedReportingConfigData', JSON.stringify( data ) );
			localStorage.setItem( 'cacheExpiry', cacheExpiryTime.toString() );

			return data;
		} else {
			throw new Error(
				`Load Reporting Config web request failed with status code ${
					response.status
				}. Response body: ${ JSON.stringify( await response.json() ) }`
			);
		}
	}

	async log( payload: LogPayload ) {
		const request = new Request( '/wp-json/bugomattic/v1/logs/', {
			method: 'POST',
			credentials: 'same-origin',
			headers: new Headers( {
				'Content-Type': 'application/json',
				[ this.nonceHeaderName ]: this.nonce,
			} ),
			body: JSON.stringify( payload ),
		} );
		const response = await fetch( request );

		if ( ! response.ok ) {
			throw new Error(
				`Log web request failed with status code ${
					response.status
				}. Response body: ${ JSON.stringify( await response.json() ) }`
			);
		}
	}

	async getRepoFilters(): Promise< string[] > {
		const request = new Request( '/wp-json/bugomattic/v1/repos/', {
			method: 'GET',
			credentials: 'same-origin',
			headers: new Headers( {
				[ this.nonceHeaderName ]: this.nonce,
			} ),
		} );
		const response = await fetch( request );

		if ( response.ok ) {
			return response.json();
		} else {
			throw new Error(
				`Get Repo Filters web request failed with status code ${
					response.status
				}. Response body: ${ JSON.stringify( await response.json() ) }`
			);
		}
	}

	async searchIssues( search: string, options?: SearchIssueOptions ): Promise< IssueApiResponse > {
		const queryParams = new URLSearchParams();
		queryParams.set( 'search', search );
		if ( options?.sort ) {
			queryParams.set( 'sort', options.sort );
		}
		if ( options?.status ) {
			queryParams.set( 'status', options.status );
		}
		if ( options?.repos ) {
			for ( const repo of options.repos ) {
				queryParams.append( 'repos[]', repo );
			}
		}

		const request = new Request( `/wp-json/bugomattic/v1/issues/?${ queryParams.toString() }`, {
			method: 'GET',
			credentials: 'same-origin',
			headers: new Headers( {
				[ this.nonceHeaderName ]: this.nonce,
			} ),
		} );
		const response = await fetch( request );

		if ( response.ok ) {
			return response.json();
		} else {
			throw new Error(
				`Search Issues web request failed with status code ${
					response.status
				}. Response body: ${ JSON.stringify( await response.json() ) }`
			);
		}
	}
}

export function createProductionApiClient(): ApiClient & LoggerApiClient {
	const productionApiClient = new ProductionApiClient();
	return {
		loadReportingConfig: productionApiClient.loadReportingConfig.bind( productionApiClient ),
		log: productionApiClient.log.bind( productionApiClient ),
		getRepoFilters: productionApiClient.getRepoFilters.bind( productionApiClient ),
		searchIssues: productionApiClient.searchIssues.bind( productionApiClient ),
	};
}
