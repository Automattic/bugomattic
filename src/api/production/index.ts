export * from './api-client';
export * from './provider';
