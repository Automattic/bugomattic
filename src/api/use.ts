import { useContext } from 'react';
import { ApiContext } from './context';

export function useApi() {
	return useContext( ApiContext );
}
