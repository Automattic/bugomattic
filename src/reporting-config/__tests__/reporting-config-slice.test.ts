import { ReportingConfigApiResponse } from '../../api';
import { reportingConfigSlice, reportingConfigReducer, createFromApiResponse } from '..';

function createFakeResponse(): ReportingConfigApiResponse {
	return {
		FakeProduct: {
			description: 'Description for fake product',
			learnMoreLinks: [
				{
					type: 'slack',
					channel: 'product-channel',
				},
			],
			taskMapping: {
				bug: [ { instructions: 'Product bug task' } ],
				featureRequest: [ { instructions: 'Product feature request task' } ],
				showStopper: [ { instructions: 'Product show stopper task' } ],
			},
			featureGroups: {
				FakeFeatureGroup: {
					description: 'Description for fake feature group',
					learnMoreLinks: [
						{
							type: 'p2',
							subdomain: 'feature-group',
						},
					],
					taskMapping: {
						bug: [ { instructions: 'Feature group bug task' } ],
						featureRequest: [ { instructions: 'Feature group request task' } ],
						showStopper: [ { instructions: 'Feature group show stopper task' } ],
					},
					features: {
						GroupNestedFeature: {
							description: 'Fake feature nested under feature group',
							keywords: [ 'foo' ],
							learnMoreLinks: [
								{
									type: 'slack',
									channel: 'feature-group-feature',
								},
							],
							taskMapping: {
								bug: [ { instructions: 'Feature nested under group bug task' } ],
								featureRequest: [ { instructions: 'Feature nested under group request task' } ],
								showStopper: [ { instructions: 'Feature nested under group show stopper task' } ],
							},
						},
					},
				},
			},
			features: {
				ProductNestedFeature: {
					description: 'Fake feature nested under product',
					keywords: [ 'bar' ],
					learnMoreLinks: [
						{
							type: 'slack',
							channel: 'product-feature',
						},
					],
					taskMapping: {
						bug: [ { instructions: 'Feature nested under product bug task' } ],
						featureRequest: [ { instructions: 'Feature nested under product request task' } ],
						showStopper: [ { instructions: 'Feature nested under product show stopper task' } ],
					},
				},
			},
		},
	};
}

describe( '[reportingConfigSlice]', () => {
	describe( 'Normalize Reporting Config Reducer', () => {
		test( 'Creates the expected normalized reporting config', () => {
			const initialState = reportingConfigSlice.getInitialState();
			const { normalized } = reportingConfigReducer(
				initialState,
				createFromApiResponse( createFakeResponse() )
			);
			expect( normalized ).toMatchSnapshot();
		} );
	} );
} );
