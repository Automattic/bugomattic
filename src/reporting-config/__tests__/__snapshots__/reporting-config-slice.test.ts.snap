// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[reportingConfigSlice] Normalize Reporting Config Reducer Creates the expected normalized reporting config 1`] = `
Object {
  "featureGroups": Object {
    "FakeProduct__FakeFeatureGroup": Object {
      "description": "Description for fake feature group",
      "id": "FakeProduct__FakeFeatureGroup",
      "learnMoreLinks": Array [
        Object {
          "subdomain": "feature-group",
          "type": "p2",
        },
      ],
      "name": "FakeFeatureGroup",
      "product": "FakeProduct",
      "taskMapping": Object {
        "bug": Array [
          Object {
            "instructions": "Feature group bug task",
          },
        ],
        "featureRequest": Array [
          Object {
            "instructions": "Feature group request task",
          },
        ],
        "showStopper": Array [
          Object {
            "instructions": "Feature group show stopper task",
          },
        ],
      },
    },
  },
  "features": Object {
    "FakeProduct__FakeFeatureGroup__GroupNestedFeature": Object {
      "description": "Fake feature nested under feature group",
      "featureGroup": "FakeProduct__FakeFeatureGroup",
      "id": "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      "keywords": Array [
        "foo",
      ],
      "learnMoreLinks": Array [
        Object {
          "channel": "feature-group-feature",
          "type": "slack",
        },
      ],
      "name": "GroupNestedFeature",
      "taskMapping": Object {
        "bug": Array [
          Object {
            "instructions": "Feature nested under group bug task",
          },
        ],
        "featureRequest": Array [
          Object {
            "instructions": "Feature nested under group request task",
          },
        ],
        "showStopper": Array [
          Object {
            "instructions": "Feature nested under group show stopper task",
          },
        ],
      },
    },
    "FakeProduct__ProductNestedFeature": Object {
      "description": "Fake feature nested under product",
      "id": "FakeProduct__ProductNestedFeature",
      "keywords": Array [
        "bar",
      ],
      "learnMoreLinks": Array [
        Object {
          "channel": "product-feature",
          "type": "slack",
        },
      ],
      "name": "ProductNestedFeature",
      "product": "FakeProduct",
      "taskMapping": Object {
        "bug": Array [
          Object {
            "instructions": "Feature nested under product bug task",
          },
        ],
        "featureRequest": Array [
          Object {
            "instructions": "Feature nested under product request task",
          },
        ],
        "showStopper": Array [
          Object {
            "instructions": "Feature nested under product show stopper task",
          },
        ],
      },
    },
  },
  "products": Object {
    "FakeProduct": Object {
      "description": "Description for fake product",
      "id": "FakeProduct",
      "learnMoreLinks": Array [
        Object {
          "channel": "product-channel",
          "type": "slack",
        },
      ],
      "name": "FakeProduct",
      "taskMapping": Object {
        "bug": Array [
          Object {
            "instructions": "Product bug task",
          },
        ],
        "featureRequest": Array [
          Object {
            "instructions": "Product feature request task",
          },
        ],
        "showStopper": Array [
          Object {
            "instructions": "Product show stopper task",
          },
        ],
      },
    },
  },
}
`;
